webapi/Skills/ChatSkills/ChatSkill.cs

```diff
@@ -394,59 +446,42 @@ private async Task<ChatMessage> GetChatResponseAsync(string chatId, string userI
         string chatHistory = string.Empty;
+        var chatHistoryTokenBudget = remainingTokenBudget - chatContextTokenCount - TokenUtilities.GetContextMessageTokenCount(AuthorRole.System, planResult);
 
-        // Append the chat history, if allowed.
-        if (chatHistoryTokenLimit > 0)
-        {
-            await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting chat history", cancellationToken);
-            chatHistory = await this.ExtractChatHistoryAsync(chatId, chatHistoryTokenLimit, cancellationToken);
-            chatContext.ThrowIfFailed();
-            chatContextText = $"{chatContextText}\n{chatHistory}";
-        }
+        // Append previous messages
+        await this.UpdateBotResponseStatusOnClientAsync(chatId, "Extracting chat history", cancellationToken);
+        chatHistory = await this.GetAllowedChatHistoryAsync(chatId, chatHistoryTokenBudget, promptTemplate, cancellationToken);
+        chatContext.ThrowIfFailed();
 
         // Append the plan result last, if exists, to imply precedence.
         if (!string.IsNullOrWhiteSpace(planResult))
         {
-            chatContextText = $"{chatContextText}\n{planResult}";
+            promptTemplate.AddSystemMessage(planResult);
         }
 
-        // Set variables needed in prompt
-        chatContext.Variables.Set("audience", audience);
-        chatContext.Variables.Set("userIntent", userIntent);
-        chatContext.Variables.Set("chatContext", chatContextText);
-
-        // Render the prompt
-        var promptRenderer = new PromptTemplateEngine();
-        var renderedPrompt = await promptRenderer.RenderAsync(
-            this._promptOptions.SystemChatPrompt,
-            chatContext,
-            cancellationToken);
-        chatContext.Variables.Set("prompt", renderedPrompt);
-
-        // Need to extract this from the rendered prompt because Time and Date are calculated during render
-        var systemChatContinuation = Regex.Match(renderedPrompt, PromptsOptions.SYSTEM_CHAT_CONTINUATION_REGEX).Value;
-        var promptView = new BotResponsePrompt(renderedPrompt, this._promptOptions.SystemDescription, this._promptOptions.SystemResponse, audience, userIntent, chatMemories, documentMemories, plannerDetails, chatHistory, systemChatContinuation);
-
-        // Calculate token usage of prompt template
-        chatContext.Variables.Set(TokenUtilities.GetFunctionKey(this._logger, "SystemMetaPrompt")!, TokenUtilities.TokenCount(renderedPrompt).ToString(CultureInfo.InvariantCulture));
-
+        var promptView = new BotResponsePrompt(this._promptOptions.SystemDescription, this._promptOptions.SystemResponse, audience, userIntent, chatMemories, documentMemories, plannerDetails, chatHistory, promptTemplate);
+        chatContext.Variables.Set(TokenUtilities.GetFunctionKey(this._logger, "SystemMetaPrompt")!, TokenUtilities.GetContextMessagesTokenCount(promptTemplate).ToString(CultureInfo.InvariantCulture));
         chatContext.ThrowIfFailed();
```

Comment:
Do we need this?

Reply:
Need what? Setting the variable in the chatContext?

Reply:
Same as above.

Reply:
removed